File: rating.py
# coding=utf-8
import json
import sys
import math


class User(object):

    def __init__(self, rank, old_rating, handle='', official_new_rating=0):
        self.rank = float(rank)
        self.old_rating = int(old_rating)
        self.seed = 1.0
        self.handle = str(handle)
        # official_new_rating: used for validating result
        self.official_new_rating = int(official_new_rating)


class RatingCalculator(object):

    def __init__(self, users):
        self.user_list = []
        for user in users:
            self.user_list.append(User(user['rank'], user['oldRating'], user['handle'], user['newRating']))

    def cal_p(self, user_a, user_b):
        return 1.0 / (1.0 + pow(10, (user_b.old_rating - user_a.old_rating) / 400.0))

    def get_ex_seed(self, user_list, rating, own_user):
        ex_user = User(0.0, rating)
        result = 1.0
        for user in user_list:
            if user != own_user:
                result += self.cal_p(user, ex_user)
        return result

    def cal_rating(self, user_list, rank, user):
        left = 1
        right = 8000
        while right - left > 1:
            mid = int((left + right) / 2)
            if self.get_ex_seed(user_list, mid, user) < rank:
                right = mid
            else:
                left = mid
        return left

    def calculate(self):
        # Calculate seed
        for i in range(len(self.user_list)):
            self.user_list[i].seed = 1.0
            for j in range(len(self.user_list)):
                if i != j:
                    self.user_list[i].seed += self.cal_p(self.user_list[j], self.user_list[i])
        # Calculate initial delta and sum_delta
        sum_delta = 0
        for user in self.user_list:
            user.delta = int(
                (self.cal_rating(self.user_list, math.sqrt(user.rank * user.seed), user) - user.old_rating) / 2)
            sum_delta += user.delta
        # Calculate first inc
        inc = int(-sum_delta / len(self.user_list)) - 1
        for user in self.user_list:
            user.delta += inc
        self.user_list = sorted(self.user_list, key=lambda x: x.old_rating, reverse=True)
        # Calculate second inc
        s = min(len(self.user_list), int(4 * round(math.sqrt(len(self.user_list)))))
        sum_s = 0
        for i in range(s):
            sum_s += self.user_list[i].delta
        inc = min(max(int(-sum_s / s), -10), 0)
        # Calculate new rating
        for user in self.user_list:
            user.delta += inc
            user.new_rating = user.old_rating + user.delta
        self.user_list = sorted(self.user_list, key=lambda x: x.rank, reverse=False)


if __name__ == '__main__':
    if len(sys.argv) < 2:
        print('Usage: python3 rating.py [codeforces_contest_id]')
        sys.exit(1)
    contest_id = sys.argv[1]
    with open('cf_rating_{}.json'.format(contest_id), 'r') as f:
        test_users = json.loads(f.read())

    # For consecutive users with same rank, we should reassign their ranks to the real rank of the last
    # For example, initial standings are [1, 2, 2, 2, 5], and new standings will be [1, 4, 4, 4, 5]
    last_idx = 0
    last_rank = 1
    for i in range(1, len(test_users)):
        if test_users[i]['rank'] > last_rank:
            for j in range(last_idx, i):
                test_users[j]['rank'] = i
            last_idx = i
            last_rank = test_users[i]['rank']
    for i in range(last_idx, len(test_users)):
        test_users[i]['rank'] = len(test_users)

    # Calculate rating changes
    calculator = RatingCalculator(test_users)
    calculator.calculate()

    res = []
    # validation: used for validating result
    validation = True
    for user in calculator.user_list:
        res.append({
            'handle': user.handle,
            'rank': int(user.rank),
            'old_rating': user.old_rating,
            'new_rating': user.new_rating,
            'rating_change': user.delta,
        })
        if user.new_rating != user.official_new_rating:
            validation = False
            print('Failed with {}. rank: {}, seed: {}, rating: {}->{} vs {}'.format(user.handle, user.rank, user.seed, user.old_rating, user.new_rating, user.official_new_rating))
    with open('cf_rating_changes_{}.json'.format(contest_id), 'w') as f:
        f.write(json.dumps(res, indent=4))
    print('Validation:', validation)

File: rating.cpp
#include <cstdio>
#include <iostream>
#include <fstream>
#include <cmath>
#include <algorithm>
#include <string>
#include <vector>

using namespace std;

struct user {
	double rank;
	string handle;
	int old_rating;
	int new_rating = 0;
	int official_new_rating;
	double seed = 1.0;
	int delta = 0;
	string validation;
} u;

vector<user> users;

// Functions for user struct

bool cmp_old_rating_desc(user a, user b) {
	return a.old_rating > b.old_rating;
}

bool cmp_rank_asc(user a, user b) {
	return a.rank < b.rank;
}

string to_string(user &u) {
	char s[200];
	sprintf(s, "%5.0f  %-24s seed: %12.6f  rating: %4d -> %4d vs %4d%s\n",
		round(u.rank),
		u.handle.c_str(),
		u.seed,
		u.old_rating,
		u.new_rating,
		u.official_new_rating,
		u.validation.c_str()
	);
	return string(s);
}

string to_string(vector<user> &users) {
	string s;
	for(int i = 0; i < users.size(); i++) {
		s += to_string(users[i]);
	}
	return s;
}

// Functions for calculation

double cal_p(user &a, user &b) {
	return 1.0 / (1.0 + pow(10, (b.old_rating - a.old_rating) / 400.0));
}

double cal_seed(int idx, int rating) {
	user ex_user;
	ex_user.old_rating = rating;
	double res = 1.0;
	for(int i = 0; i < users.size(); i++) {
		if(i != idx) {
			res += cal_p(users[i], ex_user);
		}
	}
	return res;
}

int cal_rating_to_rank(int idx, double rank) {
	int l = 1, r = 8000;
	while(r-l > 1) {
		int mid = (l + r) / 2;
		if(cal_seed(idx, mid) < rank) {
			r = mid;
		}
		else {
			l = mid;
		}
	}
	return l;
}

void work() {
	// Calculate seed
	for(int i = 0; i < users.size(); i++) {
		for(int j = 0; j < users.size(); ++j) {
			if(i != j) {
				users[i].seed += cal_p(users[j], users[i]);
			}
		}
	}
	// Calculate initial delta and sum_delta
	int sum_delta = 0;
	for(int i = 0; i < users.size(); i++) {
		double m = sqrt(users[i].rank * users[i].seed);
		int R = cal_rating_to_rank(i, m);
		users[i].delta = (R - users[i].old_rating) / 2;
		sum_delta += users[i].delta;
	}
	// Calculate the first inc
	int inc = -(sum_delta / users.size()) - 1;
	for(int i = 0; i < users.size(); i++) {
		users[i].delta += inc;
	}
	// Calculate the second inc
	sort(users.begin(), users.end(), cmp_old_rating_desc);
	int s = min((int)(users.size()), (int)(4*round(sqrt(users.size()))));
	int sum_s = 0;
	for(int i = 0; i < s; i++) {
		sum_s += users[i].delta;
	}
	inc = min(max(-(sum_s / s), -10), 0);
	for(int i = 0; i < users.size(); i++) {
		users[i].delta += inc;
	}
	// Calculate new rating
	for(int i = 0; i < users.size(); i++) {
		users[i].new_rating = users[i].old_rating + users[i].delta;
	}
	sort(users.begin(), users.end(), cmp_rank_asc);
}

// Compare calculated new rating with official new rating

int validate() {
	char s[20];
	int diff_count = 0;
	for(int i = 0; i < users.size(); i++) {
		if(users[i].new_rating != users[i].official_new_rating) {
			sprintf(s, " [diff: %5d]", users[i].new_rating - users[i].official_new_rating);
			users[i].validation = string(s);
			diff_count++;
		}
	}
	return diff_count;
}

void reassign_rank() {
	int last_idx = 0, last_rank = 1;
	for(int i = 0; i < users.size(); i++) {
		if(users[i].rank > last_rank) {
			for(int j = last_idx; j < i; ++j) {
				users[j].rank = i;
			}
			last_idx = i;
			last_rank = users[i].rank;
		}
	}
	for(int i = last_idx; i < users.size(); i++) {
		users[i].rank = users.size();
	}
}

int main(int argc, char const *argv[]) {
	if(argc < 2) {
		printf("Usage: ./file_name [codeforces_contest_id]\n");
		return 1;
	}
	string contest_id = string(argv[1]);
	string in_file = "tests/cf_rating_official_" + contest_id + ".txt";
	string out_file = "tests/cf_rating_result_" + contest_id + ".txt";
	ifstream in(in_file, ios::in);
	while(in >> u.rank >> u.handle >> u.old_rating >> u.official_new_rating) {
		users.push_back(u);
	}
	in.close();
	if(users.empty()) {
		printf("Invalid contest\n");
		return 1;
	}
	
	reassign_rank();
	work();
	int diff_count = validate();

	ofstream out(out_file, ios::out);
	out << to_string(users);
	out.close();
	printf("[Validation] diff count: %d/%d (%.2f%%)\n",
		diff_count,
		users.size(),
		100.0*diff_count/users.size()
	);
	return 0;
}